builtin/v9/verifreg/verified_registry_state.go

```diff
@@ -63,3 +75,151 @@ func ConstructState(store adt.Store, rootKeyAddress addr.Address) (*State, error
 		Claims:                   emptyMapCid,
 	}, nil
 }
+
+func (st *State) FindAllocation(store adt.Store, clientIdAddr address.Address, allocationId AllocationId) (*Allocation, bool, error) {
+	if clientIdAddr.Protocol() != address.ID {
+		return nil, false, xerrors.Errorf("can only look up ID addresses")
+	}
+
+	innerHamtCid, err := getInnerHamtCid(store, abi.IdAddrKey(clientIdAddr), st.Allocations)
+	if err != nil {
+		return nil, false, err
+	}
+
+	idToAllocationMap, err := adt.AsMap(store, innerHamtCid, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get inner map: %x", err)
+	}
+
+	var allocation Allocation
+	if found, err := idToAllocationMap.Get(allocationId, &allocation); err != nil {
+		return nil, false, xerrors.Errorf("looking up allocation ID: %d: %w", allocationId, err)
+	} else if !found {
+		return nil, false, nil
+	}
+
+	clientId, err := address.IDFromAddress(clientIdAddr)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get ID from clientIdAddr: %s", clientIdAddr)
+	}
+
+	if uint64(allocation.Client) != clientId {
+		return nil, false, xerrors.Errorf("clientId: %d did not match client in allocation: %d", clientId, allocation.Client)
+	}
```

Comment:
If the state somehow ended up like this, I don't think this is the right place to catch it. This does not mirror the Rust code.

Reply:
I'd prefer to leave it -- I think it's reasonable for the state layer to know that if it's asked to find a claim for a specific provider, then the returned claim should match that provider.

No harm done in the sanity check, I think.